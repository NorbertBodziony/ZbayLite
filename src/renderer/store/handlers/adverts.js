import BigNumber from 'bignumber.js'

import identitySelectors from '../selectors/identity'
import channelSelectors from '../selectors/channel'
import offersHandlers from '../../store/handlers/offers'
import { messageType } from '../../zbay/messages'
import { messages } from '../../zbay'
import { getClient } from '../../zcash'
import notificationsHandlers from './notifications'
import directMessagesQueueHandlers from './directMessagesQueue'
import { errorNotification, successNotification } from './utils'
import operationsHandlers, { operationTypes } from './operations'
import contactsHandlers from './contacts'

const handleSend = ({ values }) => async (dispatch, getState) => {
  const data = {
    tag: values.tag,
    background: values.background.toString(),
    title: values.title,
    provideShipping: values.shippingInfo ? '1' : '0',
    amount: values.usd,
    description: values.description
  }
  const identityAddress = identitySelectors.address(getState())
  const channel = channelSelectors.data(getState()).toJS()
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageType.AD,
      data: data,
      spent: new BigNumber('0.0001')
    },
    privKey
  })
  const transfer = await messages.messageToTransfer({
    message,
    address: channel.address,
    identityAddress
  })
  try {
    const opId = await getClient().payment.send(transfer)
    await dispatch(
      operationsHandlers.epics.observeOperation({
        opId,
        type: operationTypes.pendingMessage,
        meta: {
          channelId: channel.id,
          message: message
        }
      })
    )
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification('Advert successfully posted')
      )
    )
  } catch (err) {
    notificationsHandlers.actions.enqueueSnackbar(
      dispatch(
        errorNotification({
          message: "Couldn't send the message, please check node connection."
        })
      )
    )
  }
}

const handleSendTransfer = ({ values, history, payload }) => async (dispatch, getState) => {
  const shippingData = identitySelectors.shippingData(getState())
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageType.ITEM_TRANSFER,
      data: {
        itemId: payload.id.substring(0, 64),
        tag: payload.tag,
        offerOwner: payload.offerOwner,
        shippingData
      },
      spent: new BigNumber(values.zec)
    },
    privKey
  })
  dispatch(offersHandlers.epics.createOfferAdvert({ payload, history }))
  dispatch(directMessagesQueueHandlers.epics.addDirectMessage({
    message,
    recipientAddress: payload.address,
    recipientUsername: payload.offerOwner
  })
  )
  dispatch(contactsHandlers.epics.updateDeletedChannelTimestamp({ address: payload.id + payload.offerOwner, timestamp: 0 }))
}

export const epics = {
  handleSend,
  handleSendTransfer
}

export default {
  epics
}
