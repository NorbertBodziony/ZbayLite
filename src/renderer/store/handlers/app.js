import { ipcRenderer, remote } from 'electron'
import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import { actionTypes } from '../../../shared/static'
import { actionCreators } from './modals'

export const AppState = Immutable.Record(
  {
    version: null,
    transfers: Immutable.Map(),
    newUser: false,
    modalTabToOpen: null,
    allTransfersCount: 0,
    newTransfersCounter: 0
  },
  'AppState'
)

export const initialState = AppState()

const loadVersion = createAction(actionTypes.SET_APP_VERSION, () =>
  remote.app.getVersion()
)
const setTransfers = createAction(actionTypes.SET_TRANSFERS)
const setModalTab = createAction(actionTypes.SET_CURRENT_MODAL_TAB)
const clearModalTab = createAction(actionTypes.CLEAR_CURRENT_MODAL_TAB)
const setAllTransfersCount = createAction(actionTypes.SET_ALL_TRANSFERS_COUNT)
const setNewTransfersCount = createAction(actionTypes.SET_NEW_TRANSFERS_COUNT)
const reduceNewTransfersCount = createAction(
  actionTypes.REDUCE_NEW_TRANSFERS_COUNT
)

export const actions = {
  loadVersion,
  setTransfers,
  setModalTab,
  clearModalTab,
  setAllTransfersCount,
  setNewTransfersCount,
  reduceNewTransfersCount
}

export const askForBlockchainLocation = () => async (dispatch, getState) => {
  dispatch(actionCreators.openModal('blockchainLocation')())
}

export const proceedWithSyncing = (payload) => async (dispatch, getState) => {
  ipcRenderer.send('proceed-with-syncing', payload)
  dispatch(actionCreators.closeModal('blockchainLocation')())
}

export const reducer = handleActions(
  {
    [setNewTransfersCount]: (state, { payload: setNewTransfersCount }) =>
      state.set('newTransfersCounter', setNewTransfersCount),
    [reduceNewTransfersCount]: (state, { payload: amount }) =>
      state.update('newTransfersCounter', count => count - amount),
    [loadVersion]: (state, { payload: version }) =>
      state.set('version', version),
    [setModalTab]: (state, { payload: tabName }) =>
      state.set('modalTabToOpen', tabName),
    [setAllTransfersCount]: (state, { payload: transfersCount }) =>
      state.set('allTransfersCount', transfersCount),
    [clearModalTab]: state => state.set('modalTabToOpen', null),
    [setTransfers]: (state, { payload: { id, value } }) => {
      return state.setIn(['transfers', id], value)
    }
  },
  initialState
)

export const epics = {
  askForBlockchainLocation,
  proceedWithSyncing
}

export default {
  epics,
  actions,
  reducer
}
