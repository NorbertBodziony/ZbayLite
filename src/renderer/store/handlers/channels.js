import Immutable from 'immutable'
import { DateTime } from 'luxon'
import { createAction, handleActions } from 'redux-actions'
import BigNumber from 'bignumber.js'

import {
  typeFulfilled,
  typeRejected,
  typePending,
  errorNotification,
  successNotification,
  LoaderState
} from './utils'
import notificationsHandlers from './notifications'
import channelHandlers from './channel'
import identitySelectors from '../selectors/identity'
import channelsSelectors from '../selectors/channels'
import modalsHandlers from './modals'
import { messages } from '../../zbay'
import { getVault } from '../../vault'
import { getClient } from '../../zcash'
import { networkFee } from '../../../shared/static'

const toBigNumber = x => new BigNumber(x)

export const ChannelsState = Immutable.Record(
  {
    data: Immutable.List(),
    loader: LoaderState({ loading: true })
  },
  'ChannelsState'
)

export const initialState = ChannelsState()

export const actionTypes = {
  LOAD_CHANNELS: 'LOAD_IDENTITY_CHANNELS',
  SET_LAST_SEEN: 'SET_CHANNELS_LAST_SEEN',
  SET_UNREAD: 'SET_CHANNEL_UNREAD'
}

const loadChannels = createAction(actionTypes.LOAD_CHANNELS, async id => {
  const channels = await getVault().channels.listChannels(id)

  return channels
})
const loadChannelsToNode = createAction(actionTypes.LOAD_CHANNELS, async id => {
  const channels = await getVault().channels.listChannels(id)
  await Promise.all(
    channels.map(channel =>
      getClient().keys.importIVK({ ivk: channel.keys.ivk, address: channel.address })
    )
  )
  await Promise.all(
    channels.filter(ch => ch.keys.sk).map(channel =>
      getClient().keys.importSK({ sk: channel.keys.sk, rescan: 'no' })
    )
  )
  return channels
})

const setLastSeen = createAction(actionTypes.SET_LAST_SEEN)
const setUnread = createAction(actionTypes.SET_UNREAD)

export const actions = {
  loadChannels,
  setLastSeen,
  setUnread,
  loadChannelsToNode
}

const _createChannel = async (identityId, { name, description }) => {
  const address = await getClient().addresses.create('sapling')
  const [ivk, sk] = await Promise.all([
    getClient().keys.exportIVK(address),
    getClient().keys.exportSK(address)
  ])
  const channel = {
    address,
    name: name,
    description: description,
    private: true,
    keys: {
      ivk,
      sk
    }
  }
  await getVault().channels.importChannel(identityId, channel)
  return address
}

const createChannel = (values, formActions) => async (dispatch, getState) => {
  const closeModal = modalsHandlers.actionCreators.closeModal('createChannel')
  const balance = identitySelectors.balance('zec')(getState())
  try {
    if (balance.lt(0.0002)) {
      dispatch(
        notificationsHandlers.actions.enqueueSnackbar(
          errorNotification({ message: `You need minimum 0.0002 ZEC to create a channel. ` })
        )
      )
      formActions.setSubmitting(false)
      return
    }
    const identityId = identitySelectors.id(getState())
    const address = await _createChannel(identityId, values)
    await dispatch(channelHandlers.epics.sendChannelSettingsMessage({ address: address }))
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({ message: `Successfully created ${values.name} channel.` })
      )
    )
    formActions.setSubmitting(false)
    dispatch(closeModal())
    dispatch(loadChannels(identityId))
  } catch (error) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({ message: `Failed to create channel: ${error.message}` })
      )
    )
    formActions.setSubmitting(false)
  }
}

const withdrawMoneyFromChannels = () => async (dispatch, getState) => {
  const ownedChannels = channelsSelectors.ownedChannels(getState())
  let earnedAmount = toBigNumber(0)
  for (const channel of ownedChannels) {
    const channelAddress = channel.get('address')
    const amount = await getMoneyFromChannel(channelAddress)(dispatch, getState)
    earnedAmount = earnedAmount.plus(amount)
  }
  if (earnedAmount.gt(toBigNumber(0))) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: `You will shortly receive ${earnedAmount.toFixed(
            4
          )} zash for commission from your created channels.`
        })
      )
    )
  }
}

const getMoneyFromChannel = address => async (dispatch, getState) => {
  const amount = await getClient().accounting.balance(address)
  const identityAddress = identitySelectors.address(getState())
  if (amount.gt(0.0005)) {
    const transfer = messages.createEmptyTransfer({
      address: identityAddress,
      amount: amount.minus(networkFee).toString(),
      identityAddress: address
    })
    await getClient().payment.send(transfer)
    return amount.minus(networkFee)
  }
  return toBigNumber(0)
}

const updateLastSeen = ({ channelId }) => async (dispatch, getState) => {
  const identity = identitySelectors.data(getState())
  const lastSeen = DateTime.utc()
  await getVault().channels.updateLastSeen({
    identityId: identity.get('id'),
    channelId,
    lastSeen
  })
  dispatch(setLastSeen({ channelId, lastSeen }))
  dispatch(setUnread({ channelId, unread: 0 }))
}

export const epics = {
  createChannel,
  updateLastSeen,
  getMoneyFromChannel,
  withdrawMoneyFromChannels
}

export const reducer = handleActions(
  {
    [typePending(actionTypes.LOAD_CHANNELS)]: state =>
      state.setIn(['loader', 'loading'], true).setIn(['loader', 'message'], 'Loading channels'),
    [typeFulfilled(actionTypes.LOAD_CHANNELS)]: (state, { payload: data }) =>
      state.set('data', Immutable.fromJS(data)).setIn(['loader', 'loading'], false),
    [typeRejected(actionTypes.LOAD_CHANNELS)]: (state, { payload: error }) =>
      state.setIn(['loader', 'loading'], false),
    [setLastSeen]: (state, { payload: { channelId, lastSeen } }) => {
      const index = state.data.findIndex(channel => channel.get('id') === channelId)
      return state.updateIn(['data', index], ch => ch.set('lastSeen', lastSeen))
    },
    [setUnread]: (state, { payload: { channelId, unread } }) => {
      const index = state.data.findIndex(channel => channel.get('id') === channelId)
      return state.updateIn(['data', index], ch => ch.set('unread', unread))
    }
  },
  initialState
)

export default {
  actions,
  epics,
  reducer
}
