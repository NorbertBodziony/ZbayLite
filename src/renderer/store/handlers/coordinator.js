import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'

import channelsSelectors from '../selectors/channels'
import coordinatorSelectors from '../selectors/coordinator'
import messagesHandlers from './messages'
import contactsHandlers from './contacts'
import nodeHandlers from './node'
import identityHandlers from './identity'
import usersHandlers from './users'

export const Coordinator = Immutable.Record(
  {
    running: true
  },
  'Coordinator'
)

export const initialState = Coordinator()

export const stopCoordinator = createAction('STOP_COORDINATOR')
export const startCoordinator = createAction('START_COORDINATOR')

const actions = {
  stopCoordinator,
  startCoordinator
}
const coordinator = () => async (dispatch, getState) => {
  let index = 0
  while (coordinatorSelectors.running(getState())) {
    const actions = channelsSelectors
      .data(getState())
      .map(channel => () => messagesHandlers.epics.fetchMessages(channel))
      .push(() => contactsHandlers.epics.fetchMessages())
      .push(() => nodeHandlers.epics.getStatus())
      .push(() => identityHandlers.epics.fetchBalance())
      .push(() => usersHandlers.epics.fetchUsers())
    await dispatch(actions.get(index % actions.size)())
    index += 1
  }
}
const epics = {
  coordinator
}

export const reducer = handleActions(
  {
    [startCoordinator]: state => state.set('running', true),
    [stopCoordinator]: state => state.set('running', false)
  },
  initialState
)

export default {
  epics,
  actions,
  reducer
}
