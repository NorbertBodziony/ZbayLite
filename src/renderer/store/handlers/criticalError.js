import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import { actionTypes } from '../../../shared/static'

export const CriticalError = Immutable.Record({
  message: '',
  traceback: ''
}, 'CriticalError')

export const initialState = CriticalError()

const setCriticalError = createAction(actionTypes.SET_CRITICAL_ERROR)

export const actions = {
  setCriticalError
}

export const reducer = handleActions({
  [setCriticalError]: (state, { payload: error }) => state.merge(error)
}, initialState)

export default {
  actions,
  reducer
}
