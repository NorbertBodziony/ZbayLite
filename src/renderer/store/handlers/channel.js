import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import * as R from 'ramda'
import { createAction, handleActions } from 'redux-actions'

import operationsHandlers, { operationTypes, PendingMessageOp } from './operations'
import notificationsHandlers from './notifications'
import channelSelectors from '../selectors/channel'
import identitySelectors from '../selectors/identity'
import nodeSelectors from '../selectors/node'
import operationsSelectors from '../selectors/operations'
import { getClient } from '../../zcash'
import { messages } from '../../zbay'
import { errorNotification, LoaderState } from './utils'
import { channelToUri } from '../../zbay/channels'

export const MessagesState = Immutable.Record({
  loader: LoaderState({ loading: true }),
  data: Immutable.List(),
  errors: ''
})

export const ChannelState = Immutable.Record({
  spentFilterValue: new BigNumber(0),
  id: null,
  message: '',
  shareableUri: '',
  messages: MessagesState(),
  members: null
}, 'ChannelState')

export const initialState = ChannelState()

const setSpentFilterValue = createAction('SET_SPENT_FILTER_VALUE', (_, value) => value)
const setMessage = createAction('SET_CHANNEL_MESSAGE', R.path(['target', 'value']))
const setChannelId = createAction('SET_CHANNEL_ID')
const setMessages = createAction('SET_CHANNEL_MESSAGES')
const setLoading = createAction('SET_CHANNEL_LOADING')
const setLoadingMessage = createAction('SET_CHANNEL_LOADING_MESSAGE')
const setShareableUri = createAction('SET_CHANNEL_SHAREABLE_URI')

export const actions = {
  setLoading,
  setLoadingMessage,
  setSpentFilterValue,
  setMessage,
  setShareableUri,
  setChannelId
}

const loadMessages = () => async (dispatch, getState) => {
  const channel = channelSelectors.data(getState())
  const isTestnet = nodeSelectors.node(getState()).isTestnet
  const transfers = await getClient().payment.received(channel.get('address'))

  const pendingMessages = operationsSelectors.pendingMessages(getState())

  const msgs = await Promise.all(transfers.map(
    async (transfer) => {
      const message = await messages.transferToMessage(transfer, isTestnet)

      const pendingMessage = pendingMessages.find(
        pm => pm.txId && pm.txId === message.id)
      if (pendingMessage) {
        dispatch(
          operationsHandlers.actions.removeOperation(pendingMessage.opId)
        )
      }
      return message
    }
  ))

  dispatch(setMessages(msgs))
}

const loadChannel = (id) => async (dispatch, getState) => {
  dispatch(setLoading(true))
  dispatch(setLoadingMessage('Loading messages'))
  try {
    dispatch(setChannelId(id))

    // Calculate URI on load, that way it won't be outdated, even if someone decides
    // to update channel in vault manually
    const channel = channelSelectors.data(getState())
    const uri = await channelToUri(channel.toJS())
    dispatch(setShareableUri(uri))

    await dispatch(loadMessages())
  } catch (err) {}
  dispatch(setLoading(false))
}

const sendOnEnter = (event) => async (dispatch, getState) => {
  const enterPressed = event.nativeEvent.keyCode === 13
  const shiftPressed = event.nativeEvent.shiftKey === true
  if (enterPressed && !shiftPressed) {
    event.preventDefault()
    const message = messages.createMessage({
      identity: identitySelectors.data(getState()).toJS(),
      messageData: {
        type: messages.messageType.BASIC,
        data: event.target.value
      }
    })
    const channel = channelSelectors.data(getState()).toJS()
    const transfer = await messages.messageToTransfer({
      message,
      channel
    })
    try {
      const opId = await getClient().payment.send(transfer)
      await dispatch(operationsHandlers.epics.observeOperation({
        opId,
        type: operationTypes.pendingMessage,
        meta: PendingMessageOp({
          channelId: channel.id,
          message: Immutable.fromJS(message)
        })
      }))
    } catch (err) {
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: 'Couldn\'t send the message, please check node connection.'
        })
      )
    }
    dispatch(setMessage(''))
  }
}

const resendMessage = (message) => async (dispatch, getState) => {
  dispatch(operationsHandlers.actions.removeOperation(message.id))
  const channel = channelSelectors.data(getState()).toJS()
  const transfer = await messages.messageToTransfer({
    message,
    channel
  })
  try {
    const opId = await getClient().payment.send(transfer)
    await dispatch(operationsHandlers.epics.observeOperation({
      opId,
      type: operationTypes.pendingMessage,
      meta: PendingMessageOp({
        channelId: channel.id,
        message: Immutable.fromJS(message)
      })
    }))
  } catch (err) {
    notificationsHandlers.actions.enqueueSnackbar(
      errorNotification({
        message: 'Couldn\'t send the message, please check node connection.'
      })
    )
  }
}

export const epics = {
  sendOnEnter,
  loadChannel,
  resendMessage,
  loadMessages
}

// TODO: we should have a global loader map
export const reducer = handleActions({
  [setLoading]: (state, { payload: loading }) => state.setIn(['messages', 'loader', 'loading'], loading),
  [setLoadingMessage]: (state, { payload: message }) => state.setIn(['messages', 'loader', 'message'], message),
  [setSpentFilterValue]: (state, { payload: value }) => state.set('spentFilterValue', new BigNumber(value)),
  [setMessage]: (state, { payload: value }) => state.set('message', value),
  [setChannelId]: (state, { payload: id }) => state.set('id', id),
  [setShareableUri]: (state, { payload: uri }) => state.set('shareableUri', uri),
  [setMessages]: (state, { payload: messages }) => state.update(
    'messages',
    msgMeta => msgMeta
      .set('data', Immutable.fromJS(messages))
  )
}, initialState)

export default {
  reducer,
  epics,
  actions
}
