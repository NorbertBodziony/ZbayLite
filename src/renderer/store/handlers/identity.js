import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import { createAction, handleActions } from 'redux-actions'
import secp256k1 from 'secp256k1'
import { randomBytes } from 'crypto'

import { getClient } from '../../zcash'
import channels from '../../zcash/channels'

import identitySelectors from '../selectors/identity'
import nodeSelectors from '../selectors/node'
import appSelectors from '../selectors/app'
import channelsSelectors from '../selectors/channels'
import channelsHandlers from './channels'
import usersHandlers from './users'
import contactsHandlers from './contacts'
import messagesHandlers from './messages'
import operationHandlers, { operationTypes, ShieldBalanceOp } from './operations'
import vaultHandlers from './vault'
import ratesHandlers from './rates'
import nodeHandlers from './node'
import vault, { getVault } from '../../vault'
import migrateTo_0_2_0 from '../../../shared/migrations/0_2_0' // eslint-disable-line camelcase
import migrateTo_0_7_0 from '../../../shared/migrations/0_7_0' // eslint-disable-line camelcase
import { LoaderState } from './utils'
import modalsHandlers from './modals'

export const ShippingData = Immutable.Record(
  {
    firstName: '',
    lastName: '',
    street: '',
    country: '',
    region: '',
    city: '',
    postalCode: ''
  },
  'ShippingData'
)

const _Identity = Immutable.Record(
  {
    id: null,
    address: '',
    transparentAddress: '',
    signerPrivKey: '',
    signerPubKey: '',
    name: '',
    shippingData: ShippingData(),
    balance: null,
    lockedBalance: null,
    donationAllow: true,
    donationAddress: ''
  },
  'Identity'
)

export const Identity = values => {
  const record = _Identity(values)
  return record.set('shippingData', ShippingData(record.shippingData))
}

export const IdentityState = Immutable.Record(
  {
    data: Identity(),
    fetchingBalance: false,
    loader: LoaderState({ loading: true }),
    errors: ''
  },
  'IdentityState'
)

export const initialState = IdentityState()

export const setIdentity = createAction('SET_IDENTITY')
export const setBalance = createAction('SET_IDENTITY_BALANCE')
export const setLockedBalance = createAction('SET_IDENTITY_LOCKED_BALANCE')
export const setErrors = createAction('SET_IDENTITY_ERROR')
export const setFetchingBalance = createAction('SET_FETCHING_BALANCE')
export const setLoading = createAction('SET_IDENTITY_LOADING')
export const setLoadingMessage = createAction('SET_IDENTITY_LOADING_MESSAGE')
export const setShippingData = createAction('SET_IDENTITY_SHIPPING_DATA')
export const setDonationAllow = createAction('SET_DONATION_ALLOW')

const actions = {
  setIdentity,
  setFetchingBalance,
  setErrors,
  setLoading,
  setLoadingMessage,
  setLockedBalance,
  setShippingData,
  setBalance,
  setDonationAllow
}

export const shieldBalance = ({ from, to, amount, fee }) => async (dispatch, getState) => {
  const opId = await getClient().payment.send({
    from,
    amounts: [
      {
        address: to,
        amount: amount.toString()
      }
    ],
    fee
  })
  dispatch(
    operationHandlers.epics.observeOperation({
      opId,
      type: operationTypes.shieldBalance,
      meta: ShieldBalanceOp({
        amount,
        from,
        to
      })
    })
  )
}

export const fetchBalance = () => async (dispatch, getState) => {
  const fee = 0.0001
  dispatch(setFetchingBalance(true))
  const address = identitySelectors.address(getState())
  const tAddress = identitySelectors.transparentAddress(getState())
  try {
    const balance = await getClient().accounting.balance(address)
    const transparentBalance = await getClient().accounting.balance(tAddress)
    const realTBalance = transparentBalance.minus(fee)
    const lockedBalanceNotes = await getClient().payment.unspentNotes({
      addresses: [address],
      minConfirmations: 0,
      maxConfirmations: 0
    })
    const lockedBalance = lockedBalanceNotes.reduce(
      (acc, current) => acc.plus(new BigNumber(current.amount || 0)),
      new BigNumber(0)
    )
    if (realTBalance.gt(0)) {
      await dispatch(
        shieldBalance({
          from: tAddress,
          to: address,
          amount: realTBalance,
          fee
        })
      )
    }
    dispatch(setLockedBalance(lockedBalance))
    dispatch(setBalance(balance))
  } catch (err) {
    dispatch(setErrors(err.message))
  } finally {
    dispatch(setFetchingBalance(false))
  }
}

export const createSignerKeys = () => {
  let signerPrivKey
  do {
    signerPrivKey = randomBytes(32)
  } while (!secp256k1.privateKeyVerify(signerPrivKey))
  return {
    signerPrivKey: signerPrivKey.toString('hex'),
    signerPubKey: secp256k1.publicKeyCreate(signerPrivKey, true).toString('hex')
  }
}

export const createIdentity = ({ name }) => async (dispatch, getState) => {
  try {
    const [{ value: address }, transparentAddress] = await Promise.all([
      dispatch(nodeHandlers.actions.createAddress()),
      getClient().addresses.createTransparent()
    ])
    const [tpk, sk] = await Promise.all([
      await getClient().keys.exportTPK(transparentAddress),
      await getClient().keys.exportSK(address)
    ])

    const { signerPrivKey, signerPubKey } = exportFunctions.createSignerKeys()

    const { value: identity } = await dispatch(
      vaultHandlers.actions.createIdentity({
        name,
        address,
        transparentAddress,
        signerPubKey,
        signerPrivKey,
        keys: {
          tpk,
          sk
        }
      })
    )

    const network = nodeSelectors.network(getState())
    const generalChannel = channels.general[network]
    const usersChannel = channels.registeredUsers[network]
    await getVault().channels.importChannel(identity.id, generalChannel)
    await getVault().channels.importChannel(identity.id, usersChannel)
    await getClient().keys.importIVK({
      ivk: generalChannel.keys.ivk,
      address: generalChannel.address
    })
    return identity
  } catch (err) {
    dispatch(setErrors(err))
  }
}

export const setIdentityEpic = identityToSet => async (dispatch, getState) => {
  dispatch(setLoading(true))
  try {
    dispatch(setLoadingMessage('Ensuring identity integrity'))
    let identity = await migrateTo_0_2_0.ensureIdentityHasKeys(identityToSet)
    // Make sure identity is handled by the node
    dispatch(setLoadingMessage('Ensuring node contains identity keys'))
    await Promise.all([
      getClient().keys.importSK({ sk: identity.keys.sk }),
      getClient().keys.importTPK(identity.keys.tpk)
    ])

    dispatch(setLoadingMessage('Setting identity'))
    // Check if identity has signerKeys
    if (!identity.signerPrivKey || !identity.signerPubKey) {
      const { signerPrivKey, signerPubKey } = createSignerKeys()
      const { value: updatedIdentity } = await dispatch(
        vaultHandlers.actions.updateIdentitySignerKeys({
          id: identity.id,
          signerPubKey,
          signerPrivKey
        })
      )
      identity = updatedIdentity
    }
    await dispatch(setIdentity(identity))
    const network = nodeSelectors.network(getState())
    await migrateTo_0_7_0.ensureDefaultChannels(identity, network)
    dispatch(setLoadingMessage('Fetching balance and loading channels'))
    dispatch(ratesHandlers.epics.fetchPrices())
    await Promise.all([
      dispatch(fetchBalance()),
      dispatch(channelsHandlers.actions.loadChannels(identity.id))
    ])
    dispatch(setLoadingMessage('Loading users and messages'))
    await dispatch(usersHandlers.epics.fetchUsers())
    await Promise.all([
      dispatch(contactsHandlers.epics.fetchMessages()),
      dispatch(contactsHandlers.epics.loadAllSentMessages())
    ])
    await Promise.all(
      channelsSelectors
        .data(getState())
        .map(channel => dispatch(messagesHandlers.epics.fetchMessages(channel)))
    )
  } catch (err) {}
  dispatch(setLoading(false))

  const balance = identitySelectors.balance('zec')(getState())
  const lockedBalance = identitySelectors.lockedBalance('zec')(getState())
  const newUser = appSelectors.newUser(getState())
  if (lockedBalance.plus(balance).lt(0.0002) && newUser === false) {
    setTimeout(() => dispatch(modalsHandlers.actionCreators.openModal('depositMoney')()), 500)
  }
}

export const updateShippingData = (values, formActions) => async (dispatch, getState) => {
  const id = identitySelectors.id(getState())
  const identity = await vault.identity.updateShippingData(id, values)
  await dispatch(setShippingData(identity.shippingData))
  formActions.setSubmitting(false)
}
export const updateDonation = allow => async (dispatch, getState) => {
  const id = identitySelectors.id(getState())
  const identity = await vault.identity.updateDonation(id, allow)
  await dispatch(setDonationAllow(identity.donationAllow))
}

const epics = {
  fetchBalance,
  createIdentity,
  setIdentity: setIdentityEpic,
  updateShippingData,
  createSignerKeys,
  updateDonation
}

const exportFunctions = {
  createSignerKeys
}

export const reducer = handleActions(
  {
    [setLoading]: (state, { payload: loading }) => state.setIn(['loader', 'loading'], loading),
    [setLoadingMessage]: (state, { payload: message }) =>
      state.setIn(['loader', 'message'], message),
    [setIdentity]: (state, { payload: identity }) =>
      state.update('data', data => data.merge(identity)),
    [setBalance]: (state, { payload: balance }) =>
      state.update('data', data => data.set('balance', balance)),
    [setLockedBalance]: (state, { payload: balance }) =>
      state.update('data', data => data.set('lockedBalance', balance)),
    [setFetchingBalance]: (state, { payload: fetching }) => state.set('fetchingBalance', fetching),
    [setErrors]: (state, { payload: errors }) => state.set('errors', errors),
    [setShippingData]: (state, { payload: shippingData }) =>
      state.setIn(['data', 'shippingData'], ShippingData(shippingData)),
    [setDonationAllow]: (state, { payload: allow }) =>
      state.setIn(['data', 'donationAllow'], allow)
  },
  initialState
)

export default {
  actions,
  epics,
  reducer,
  exportFunctions
}
