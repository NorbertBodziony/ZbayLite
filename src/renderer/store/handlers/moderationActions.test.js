/* eslint import/first: 0 */

jest.mock('../../zcash')

import Immutable from 'immutable'
import * as R from 'ramda'
import { DateTime } from 'luxon'

import create from '../create'
import { createChannel } from '../../testUtils'
import { mock as zcashMock } from '../../zcash'
import { IdentityState, Identity } from './identity'
import { ChannelState } from '../handlers/channel'
import { ChannelsState } from '../handlers/channels'
import moderationActionsHandler from './moderationActions'
import { moderationActionsType } from '../../../shared/static'

describe('epics', () => {
  const store = create({
    initialState: Immutable.Map({
      identity: IdentityState({
        data: Identity({
          id: 'test-id',
          address: 'test address',
          name: 'Saturn',
          balance: '33.583004',
          signerPrivKey: '879aff43df53606d8ae1219d9347360e7a30d1c2f141e14c9bc96bb29bf930cb'
        })
      }),
      channel: ChannelState({
        spentFilterValue: 38,
        id: 'test',
        address: 'testAddress',
        shareableUri: 'test://uri',
        members: 2,
        message: 'Message written in the input'
      }),
      channels: ChannelsState({
        data: Immutable.fromJS([createChannel('test')])
      })
    })
  })
  jest.spyOn(DateTime, 'utc').mockImplementation(() => {
    return DateTime.fromSeconds(123)
  })
  it('sends moderation action', async () => {
    const actionPromise = store.dispatch(
      moderationActionsHandler.epics.handleModerationAction({
        moderationType: moderationActionsType.REMOVE_MESSAGE,
        moderationTarget: 'test-message-id'
      })
    )

    await actionPromise
    const result = zcashMock.requestManager.z_sendmany.mock.calls
    expect(
      R.sortBy(call => {
        const [, transfer] = call
        return transfer.address
      }, result)
    ).toMatchSnapshot()
  })
})
