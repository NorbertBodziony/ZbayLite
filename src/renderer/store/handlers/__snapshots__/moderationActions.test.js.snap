// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`epics sends moderation action 1`] = `
Array [
  Array [
    "test address",
    Array [
      Object {
        "address": "zs1testaddresstest",
        "amount": "0",
        "memo": "ff0776260f52538a895b8120df7d7c81bd387bd18851bfb8ac6f6e5fe7a77d95087e1d00b198a6d0d196bc1bd53faa0ae8f9e21af3a5428ae098101a21a1341f9b61000000007b52454d4f56455f4d45535341474500000000000040746573742d6d6573736167652d6964000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      },
    ],
    1,
    0.0000025,
  ],
]
`;
